Move implicit mapping out of CSSPrimitiveValue

darinadler opened the pull request:
#### 3f84cb0e5e030791bc3c6efbd7af3bfcca73b3b7
<pre>
Move implicit mapping out of CSSPrimitiveValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=251567">https://bugs.webkit.org/show_bug.cgi?id=251567</a>
rdar://104946191

Reviewed by Tim Nguyen.

Our automatically generated style building code depends on implicitly
converting CSSValue objects to various destination types. That was done
by a conversion operator in CSSPrimitiveValue, but it&apos;s better if the
conversion operator is separate, so the conversion can&apos;t be done by
accident in normal uses of CSSPrimitiveValue. At the very least, such
unintentional conversions can lead to confusing compiler errors.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::setDisplay): Use fromCSSValue.

* Source/WebCore/css/CSSPrimitiveValue.h: Deleted all the conversion
operators, including the ones for numeric types, and the template.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::fromCSSValue): Added.
(WebCore::TypeDeducingCSSValueMapper::TypeDeducingCSSValueMapper): Added.
(WebCore::TypeDeducingCSSValueMapper::operator TargetType const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator const CSSPrimitiveValue&amp; const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator unsigned short const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator int const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator unsigned const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator float const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator double const): Added.
(WebCore::TypeDeducingCSSValueMapper::numericValue const): Added.
(WebCore::fromCSSValueDeducingType): Added.
(WebCore::fromCSSValue): Replaced CSSPrimitiveValue specialization for LineClampValue with
fromCSSValue specialization.
(WebCore::CSSPrimitiveValue::operator ColumnSpan const): Deleted. This was not needed because we
no longer allow a number for column-span; when we removed support in the parser we missed this.
(WebCore::fromCSSValueID): Remove unneeded specializations for OptionSet&lt;HangingPunctuation&gt;,
OptionSet&lt;TextDecorationLine&gt;, and OptionSet&lt;TouchAction&gt;, using
DEFINE_TO_FROM_CSS_VALUE_ID_FUNCTIONS instead.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillClip): Use fromCSSValue.
(WebCore::CSSToStyleMap::mapFillComposite): Ditto.
(WebCore::CSSToStyleMap::mapFillBlendMode): Ditto.
(WebCore::CSSToStyleMap::mapFillOrigin): Ditto.
(WebCore::CSSToStyleMap::mapFillXPosition): Ditto.
(WebCore::CSSToStyleMap::mapFillYPosition): Ditto.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h: Updated since the underlying
type for BoxOrient is bool instead of uint8_t.

* Source/WebCore/css/process-css-properties.py:
(GenerationContext.generate_heading): Tweak comment.
(GenerateCSSPropertyNames): Use fromCSSValueDeducingType.

* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::setTextAlignmentForChangedBaseWritingDirection): Use propertyAsValueID,
CSSValueID, and nameString.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const): Use propertyAsValueID
and fromCSSValueID.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;): Removed support for HangingPunctuation::None.

* Source/WebCore/rendering/style/RenderStyleConstants.h: Converted all two-value enumerations
to be based on bool instead of uint8_t. Removed HangingPunctuation::None.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextDecorationLine): Use fromCSSValue.
(WebCore::Style::BuilderConverter::convertTextAlign): Ditto.
(WebCore::Style::BuilderConverter::convertTextAlignLast): Ditto.
(WebCore::Style::BuilderConverter::convertPathOperation): Ditto.
(WebCore::Style::BuilderConverter::convertResize): Ditto.
(WebCore::Style::BuilderConverter::convertTextUnderlinePosition): Ditto.
(WebCore::Style::BuilderConverter::convertReflection): Ditto.
(WebCore::Style::BuilderConverter::convertShapeValue): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapType): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapAlign): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapStop): Ditto.
(WebCore::Style::BuilderConverter::convertTouchAction): Ditto.
(WebCore::Style::BuilderConverter::convertSelfOrDefaultAlignmentData): Ditto.
(WebCore::Style::BuilderConverter::convertSpeakAs): Ditto.
(WebCore::Style::BuilderConverter::convertHangingPunctuation): Ditto.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueDirection): Ditto.
(WebCore::Style::BuilderCustom::applyValueVerticalAlign): Ditto.
(WebCore::Style::BuilderCustom::applyValueListStyleType): Ditto.
(WebCore::Style::BuilderCustom::applyValueOutlineStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueWritingMode): Ditto.
(WebCore::Style::BuilderCustom::applyValueTextOrientation): Ditto.
(WebCore::Style::BuilderCustom::applyValueDisplay): Ditto.
(WebCore::Style::BuilderCustom::applyValueTextEmphasisStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueCursor): Ditto.

* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::colorInterpolation const): Ditto.

Canonical link: <a href="https://commits.webkit.org/259871@main">https://commits.webkit.org/259871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d182c81af891b767565c56e55eb3d7811a5ed745

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106242 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15296 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115432 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175531 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110150 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6512 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98450 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115114 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112004 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12728 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95704 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40273 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27347 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81961 "Found 1 new API test failure: /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/cut-copy-paste/editable (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8537 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28699 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9044 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5273 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14659 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48245 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6823 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10580 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
r=me assuming crash is fixed.

darinadler committed:
Move implicit mapping out of CSSPrimitiveValue
https://bugs.webkit.org/show_bug.cgi?id=251567
rdar://104946191

Reviewed by Tim Nguyen.

Our automatically generated style building code depends on implicitly
converting CSSValue objects to various destination types. That was done
by a conversion operator in CSSPrimitiveValue, but it's better if the
conversion operator is separate, so the conversion can't be done by
accident in normal uses of CSSPrimitiveValue. At the very least, such
unintentional conversions can lead to confusing compiler errors.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::setDisplay): Use fromCSSValue.

* Source/WebCore/css/CSSPrimitiveValue.h: Deleted all the conversion
operators, including the ones for numeric types, and the template.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::fromCSSValue): Added.
(WebCore::TypeDeducingCSSValueMapper::TypeDeducingCSSValueMapper): Added.
(WebCore::TypeDeducingCSSValueMapper::operator TargetType const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator const CSSPrimitiveValue& const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator unsigned short const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator int const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator unsigned const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator float const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator double const): Added.
(WebCore::TypeDeducingCSSValueMapper::numericValue const): Added.
(WebCore::fromCSSValueDeducingType): Added.
(WebCore::fromCSSValue): Replaced CSSPrimitiveValue specialization for LineClampValue with
fromCSSValue specialization.
(WebCore::CSSPrimitiveValue::operator ColumnSpan const): Deleted. This was not needed because we
no longer allow a number for column-span; when we removed support in the parser we missed this.
(WebCore::fromCSSValueID): Remove unneeded specializations for OptionSet<HangingPunctuation>,
OptionSet<TextDecorationLine>, and OptionSet<TouchAction>, using
DEFINE_TO_FROM_CSS_VALUE_ID_FUNCTIONS instead.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillClip): Use fromCSSValue.
(WebCore::CSSToStyleMap::mapFillComposite): Ditto.
(WebCore::CSSToStyleMap::mapFillBlendMode): Ditto.
(WebCore::CSSToStyleMap::mapFillOrigin): Ditto.
(WebCore::CSSToStyleMap::mapFillXPosition): Ditto.
(WebCore::CSSToStyleMap::mapFillYPosition): Ditto.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h: Updated since the underlying
type for BoxOrient is bool instead of uint8_t.

* Source/WebCore/css/process-css-properties.py:
(GenerationContext.generate_heading): Tweak comment.
(GenerateCSSPropertyNames): Use fromCSSValueDeducingType.

* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::setTextAlignmentForChangedBaseWritingDirection): Use propertyAsValueID,
CSSValueID, and nameString.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const): Use propertyAsValueID
and fromCSSValueID.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<): Removed support for HangingPunctuation::None.

* Source/WebCore/rendering/style/RenderStyleConstants.h: Converted all two-value enumerations
to be based on bool instead of uint8_t. Removed HangingPunctuation::None.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextDecorationLine): Use fromCSSValue.
(WebCore::Style::BuilderConverter::convertTextAlign): Ditto.
(WebCore::Style::BuilderConverter::convertTextAlignLast): Ditto.
(WebCore::Style::BuilderConverter::convertPathOperation): Ditto.
(WebCore::Style::BuilderConverter::convertResize): Ditto.
(WebCore::Style::BuilderConverter::convertTextUnderlinePosition): Ditto.
(WebCore::Style::BuilderConverter::convertReflection): Ditto.
(WebCore::Style::BuilderConverter::convertShapeValue): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapType): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapAlign): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapStop): Ditto.
(WebCore::Style::BuilderConverter::convertTouchAction): Ditto.
(WebCore::Style::BuilderConverter::convertSelfOrDefaultAlignmentData): Ditto.
(WebCore::Style::BuilderConverter::convertSpeakAs): Ditto.
(WebCore::Style::BuilderConverter::convertHangingPunctuation): Ditto.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueDirection): Ditto.
(WebCore::Style::BuilderCustom::applyValueVerticalAlign): Ditto.
(WebCore::Style::BuilderCustom::applyValueListStyleType): Ditto.
(WebCore::Style::BuilderCustom::applyValueOutlineStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueWritingMode): Ditto.
(WebCore::Style::BuilderCustom::applyValueTextOrientation): Ditto.
(WebCore::Style::BuilderCustom::applyValueDisplay): Ditto.
(WebCore::Style::BuilderCustom::applyValueTextEmphasisStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueCursor): Ditto.

* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::colorInterpolation const): Ditto.

Canonical link: https://commits.webkit.org/259871@main
